File: BaekJoon/백트래킹/14888_연산자 끼워넣기.kt
import java.lang.Integer.max
import java.lang.Integer.min
import java.util.*

var n = 0
var max_value = Int.MIN_VALUE
var min_value = Int.MAX_VALUE
lateinit var array: Array<Int>
lateinit var operator: Array<Int>

fun main() = with(Scanner(System.`in`)) {
    n = nextInt()
    array = Array(n) { nextInt() }
    operator = Array(4) { nextInt() }

    calculate(array[0], 1)
    println(max_value)
    println(min_value)
}

fun calculate(sum: Int, index: Int) {
    if (index == n) {
        max_value = max(sum, max_value)
        min_value = min(sum, min_value)
        return
    }

    for (i in 0..3) {
        if (operator[i] > 0) {
            operator[i]--
            when (i) {
                0 -> calculate(sum + array[index], index + 1)
                1 -> calculate(sum - array[index], index + 1)
                2 -> calculate(sum * array[index], index + 1)
                3 -> calculate(sum / array[index], index + 1)
            }
            operator[i]++
        }
    }
}

File: BaekJoon/백트래킹/14888_연산자 끼워넣기.swift
import Foundation

let n = Int(readLine()!)!
let numbers = readLine()!.split(separator: " ").map { Int(String($0))!}
var ops = readLine()!.split(separator: " ").map { Int(String($0))!}

var maxValue = Int.min
var minValue = Int.max

calculate(numbers[0], 1)
print(maxValue)
print(minValue)

func calculate(_ sum: Int, _ depth: Int) {
    if depth == n {
        maxValue = max(maxValue, sum)
        minValue = min(minValue, sum)
        return
    }
    
    for i in 0..<4 {
        if ops[i] > 0 {
            ops[i] -= 1
            switch i {
            case 0:
                calculate(sum + numbers[depth], depth + 1)
            case 1:
                calculate(sum - numbers[depth], depth + 1)
            case 2:
                calculate(sum * numbers[depth], depth + 1)
            default:
                calculate(sum / numbers[depth], depth + 1)
            }
            ops[i] += 1
        }
    }
}
